server/src/utils.ts

```diff
@@ -74,21 +76,72 @@ export let findProjectRootOfFile = (
   }
 };
 
-// Check if binaryName exists inside binaryDirPath and return the joined path.
-export let findBinary = (
-  binaryDirPath: p.DocumentUri | null,
-  binaryName: string
-): p.DocumentUri | null => {
-  if (binaryDirPath == null) {
+// If ReScript < 12.0.0-alpha.13, then we want `{project_root}/node_modules/rescript/{c.platformDir}/{binary}`.
+// Otherwise, we want to dynamically import `{project_root}/node_modules/rescript` and from `binPaths` get the relevant binary.
+// We won't know which version is in the project root until we read and parse `{project_root}/node_modules/rescript/package.json`
+let findBinary = async (
+  projectRootPath: p.DocumentUri | null,
+  binary: "bsc.exe" | "rescript-editor-analysis.exe" | "rescript"
+) => {
+  if (config.extensionConfiguration.platformPath != null) {
+    return path.join(config.extensionConfiguration.platformPath, binary);
+  }
+
+  const rescriptDir = lookup.findFilePathFromProjectRoot(
+    projectRootPath,
+    path.join("node_modules", "rescript")
+  );
+  if (rescriptDir == null) {
     return null;
   }
-  let binaryPath: p.DocumentUri = path.join(binaryDirPath, binaryName);
-  if (fs.existsSync(binaryPath)) {
-    return binaryPath;
+
+  let rescriptVersion = null;
+  let rescriptJSWrapperPath = null
+  try {
+    const rescriptPackageJSONPath = path.join(rescriptDir, "package.json");
+    const rescriptPackageJSON = JSON.parse(await fsAsync.readFile(rescriptPackageJSONPath, "utf-8"));
+    rescriptVersion = rescriptPackageJSON.version
+    rescriptJSWrapperPath = rescriptPackageJSON.bin.rescript
+  } catch (error) {
+    return null
+  }
+
+  let binaryPath: string | null = null
+  if (binary == "rescript") {
+    // Can't use the native bsb/rescript since we might need the watcher -w
+    // flag, which is only in the JS wrapper
+    binaryPath = path.join(rescriptDir, rescriptJSWrapperPath)
+  } else if (semver.gte(rescriptVersion, "12.0.0-alpha.13")) {
+    // TODO: export `binPaths` from `rescript` package so that we don't need to
+    // copy the logic for figuring out `target`.
+    const target = `${process.platform}-${process.arch}`;
+    const targetPackagePath = path.join(rescriptDir, "..", `@rescript/${target}/bin.js`)
+    const { binPaths } = await import(targetPackagePath);
```

Comment:
Unfortunately I couldn't just do `import(rescriptDir)` to get bin paths as the `rescript` package only exports `./lib/` and `./package.json` in the `exports` field of its `package.json`.

We probably either want to export `./common/bins.js` or have a default export that reexports `binPaths`.

As for `@rescript/${target}`, I had to append `/bin.js` to the import path for it to work as, like `rescript`, the per-platform binary packages have no default export.

Reply:
Oh, you're right. I didn't re-export it because I wasn't sure the API from the compiler (it will be the first programmatic API, so...)

Reply:
> As for @rescript/${target}, I had to append /bin.js to the import path for it to work as, like rescript, the per-platform binary packages have no default export.

There is the `exports` field (which always has precedence over the `main` field). Importing the binary package without the `/bin.js` suffix should also work.

Reply:
> Don't forget to use the right compilerOptions.nodeResolution option in the tsconfig.json. Use NodeNext as long as possible, or Bundler.

Ah I didn't realise we were bundling as CJS, this explains the error messages I was getting when trying to import without `bin.js`. I'll have a go at tweaking `compilerOptions.nodeResolution` to make this work :+1:

Reply:
I tried the following:

- `compilerOptions.moduleResolution` values `NodeNext` and `Bundler`
- Updating typescript to v5.8.3
- ES2022 instead of ES1029

but I still can't get it to work :(

Reply:
This is still WIP, right? Or could we merge this and do the above in a follow up?

Reply:
We can improve the dynamic import path in a followup PR 👍